[CI][test_plugins.py] `build_and_deploy` finishes too early, before grafana-agent is done setting itself up

Some CI runs are failing in `test_plugins.py` in `test_monitoring_user_fetch_prometheus_data`. [CI run example](https://github.com/canonical/opensearch-operator/actions/runs/11310272526/job/31455221638)

Error: 
```
AssertionError: assert 0 > 500
 +  where 0 = <built-in method count of str object at 0x79c198ced030>('opensearch_')
 +    where <built-in method count of str object at 0x79c198ced030> = 'Unauthorized'.count
```

Status:
```
Model  Controller           Cloud/Region         Version  SLA          Timestamp
test   localhost-localhost  localhost/localhost  3.[5](https://github.com/canonical/opensearch-operator/actions/runs/11310272526/job/31455221638#step:37:6).3    unsupported  03:32:11Z

App                       Version  Status   Scale  Charm                     Channel        Rev  Exposed  Message
failover                           active       2  opensearch                                 0  no       
grafana-agent                      blocked      4  grafana-agent             latest/edge    28[6](https://github.com/canonical/opensearch-operator/actions/runs/11310272526/job/31455221638#step:37:7)  no       Missing ['grafana-cloud-config']|['grafana-dashboards-provider']|['logging-consumer']|['send-remote-write'] for cos-a...
main                               active       1  opensearch                                 1  no       
opensearch                         active       1  opensearch                                 2  no       
self-signed-certificates           active       1  self-signed-certificates  latest/stable  155  no       

Unit                         Workload     Agent      Machine  Public address  Ports     Message
failover/0                   active       idle       3        10.64.150.109   9200/tcp  
  grafana-agent/3            maintenance  executing           10.64.150.109             (install) Installing grafana-agent snap
failover/1*                  active       idle       4        10.64.150.206   9200/tcp  
  grafana-agent/1*           maintenance  executing           10.64.150.206             (install) Installing grafana-agent snap
main/0*                      active       idle       1        10.64.150.1[8](https://github.com/canonical/opensearch-operator/actions/runs/11310272526/job/31455221638#step:37:9)5   9200/tcp  
  grafana-agent/2            blocked      idle                10.64.150.185             Missing ['grafana-cloud-config']|['grafana-dashboards-provider']|['logging-consumer']|['send-remote-write'] for cos-a...
opensearch/0*                active       idle       2        10.64.150.200   [9](https://github.com/canonical/opensearch-operator/actions/runs/11310272526/job/31455221638#step:37:10)200/tcp  
  grafana-agent/4            blocked      idle                10.64.150.200             Missing ['grafana-cloud-config']|['grafana-dashboards-provider']|['logging-consumer']|['send-remote-write'] for cos-a...
self-signed-certificates/0*  active       idle       0        [10](https://github.com/canonical/opensearch-operator/actions/runs/11310272526/job/31455221638#step:37:11).64.150.154             

Machine  State    Address        Inst id        Base          AZ  Message
0        started  10.64.150.154  juju-9c62cd-0  ubuntu@22.04      Running
1        started  10.64.150.185  juju-9c62cd-1  ubuntu@22.04      Running
2        started  10.64.[15](https://github.com/canonical/opensearch-operator/actions/runs/11310272526/job/31455221638#step:37:16)0.200  juju-9c62cd-2  ubuntu@22.04      Running
3        started  10.64.150.109  juju-9c62cd-3  ubuntu@22.04      Running
4        started  10.64.150.206  juju-9c62cd-4  ubuntu@22.04      Running
```

The `wait_until` [here](https://github.com/canonical/opensearch-operator/blob/2/edge/tests/integration/plugins/test_plugins.py#L91) must also wait for the grafana-agent.

---------------------------------------------------------------------------------------------------

Full error stack:
```
Traceback (most recent call last):
  File "/home/ubuntu/actions-runner/_work/opensearch-operator/opensearch-operator/.tox/integration/lib/python3.10/site-packages/_pytest/runner.py", line 341, in from_call
    result: TResult | None = func()
  File "/home/ubuntu/actions-runner/_work/opensearch-operator/opensearch-operator/.tox/integration/lib/python3.10/site-packages/_pytest/runner.py", line 242, in <lambda>
    lambda: runtest_hook(item=item, **kwds), when=when, reraise=reraise
  File "/home/ubuntu/actions-runner/_work/opensearch-operator/opensearch-operator/.tox/integration/lib/python3.10/site-packages/pluggy/_hooks.py", line 513, in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
  File "/home/ubuntu/actions-runner/_work/opensearch-operator/opensearch-operator/.tox/integration/lib/python3.10/site-packages/pluggy/_manager.py", line 120, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "/home/ubuntu/actions-runner/_work/opensearch-operator/opensearch-operator/.tox/integration/lib/python3.10/site-packages/pluggy/_callers.py", line 182, in _multicall
    return outcome.get_result()
  File "/home/ubuntu/actions-runner/_work/opensearch-operator/opensearch-operator/.tox/integration/lib/python3.10/site-packages/pluggy/_result.py", line 100, in get_result
    raise exc.with_traceback(exc.__traceback__)
  File "/home/ubuntu/actions-runner/_work/opensearch-operator/opensearch-operator/.tox/integration/lib/python3.10/site-packages/pluggy/_callers.py", line 167, in _multicall
    teardown.throw(outcome._exception)
  File "/home/ubuntu/actions-runner/_work/opensearch-operator/opensearch-operator/.tox/integration/lib/python3.10/site-packages/_pytest/threadexception.py", line 92, in pytest_runtest_call
    yield from thread_exception_runtest_hook()
  File "/home/ubuntu/actions-runner/_work/opensearch-operator/opensearch-operator/.tox/integration/lib/python3.10/site-packages/_pytest/threadexception.py", line 68, in thread_exception_runtest_hook
    yield
  File "/home/ubuntu/actions-runner/_work/opensearch-operator/opensearch-operator/.tox/integration/lib/python3.10/site-packages/pluggy/_callers.py", line 167, in _multicall
    teardown.throw(outcome._exception)
  File "/home/ubuntu/actions-runner/_work/opensearch-operator/opensearch-operator/.tox/integration/lib/python3.10/site-packages/_pytest/unraisableexception.py", line 95, in pytest_runtest_call
    yield from unraisable_exception_runtest_hook()
  File "/home/ubuntu/actions-runner/_work/opensearch-operator/opensearch-operator/.tox/integration/lib/python3.10/site-packages/_pytest/unraisableexception.py", line 70, in unraisable_exception_runtest_hook
    yield
  File "/home/ubuntu/actions-runner/_work/opensearch-operator/opensearch-operator/.tox/integration/lib/python3.10/site-packages/pluggy/_callers.py", line 167, in _multicall
    teardown.throw(outcome._exception)
  File "/home/ubuntu/actions-runner/_work/opensearch-operator/opensearch-operator/.tox/integration/lib/python3.10/site-packages/_pytest/logging.py", line 846, in pytest_runtest_call
    yield from self._runtest_for(item, "call")
  File "/home/ubuntu/actions-runner/_work/opensearch-operator/opensearch-operator/.tox/integration/lib/python3.10/site-packages/_pytest/logging.py", line 829, in _runtest_for
    yield
  File "/home/ubuntu/actions-runner/_work/opensearch-operator/opensearch-operator/.tox/integration/lib/python3.10/site-packages/pluggy/_callers.py", line 167, in _multicall
    teardown.throw(outcome._exception)
  File "/home/ubuntu/actions-runner/_work/opensearch-operator/opensearch-operator/.tox/integration/lib/python3.10/site-packages/_pytest/capture.py", line 880, in pytest_runtest_call
    return (yield)
  File "/home/ubuntu/actions-runner/_work/opensearch-operator/opensearch-operator/.tox/integration/lib/python3.10/site-packages/pluggy/_callers.py", line 167, in _multicall
    teardown.throw(outcome._exception)
  File "/home/ubuntu/actions-runner/_work/opensearch-operator/opensearch-operator/.tox/integration/lib/python3.10/site-packages/_pytest/skipping.py", line 257, in pytest_runtest_call
    return (yield)
  File "/home/ubuntu/actions-runner/_work/opensearch-operator/opensearch-operator/.tox/integration/lib/python3.10/site-packages/pluggy/_callers.py", line 103, in _multicall
    res = hook_impl.function(*args)
  File "/home/ubuntu/actions-runner/_work/opensearch-operator/opensearch-operator/.tox/integration/lib/python3.10/site-packages/_pytest/runner.py", line 174, in pytest_runtest_call
    item.runtest()
  File "/home/ubuntu/actions-runner/_work/opensearch-operator/opensearch-operator/.tox/integration/lib/python3.10/site-packages/_pytest/python.py", line 1627, in runtest
    self.ihook.pytest_pyfunc_call(pyfuncitem=self)
  File "/home/ubuntu/actions-runner/_work/opensearch-operator/opensearch-operator/.tox/integration/lib/python3.10/site-packages/pluggy/_hooks.py", line 513, in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
  File "/home/ubuntu/actions-runner/_work/opensearch-operator/opensearch-operator/.tox/integration/lib/python3.10/site-packages/pluggy/_manager.py", line 120, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
  File "/home/ubuntu/actions-runner/_work/opensearch-operator/opensearch-operator/.tox/integration/lib/python3.10/site-packages/pluggy/_callers.py", line 182, in _multicall
    return outcome.get_result()
  File "/home/ubuntu/actions-runner/_work/opensearch-operator/opensearch-operator/.tox/integration/lib/python3.10/site-packages/pluggy/_result.py", line 100, in get_result
    raise exc.with_traceback(exc.__traceback__)
  File "/home/ubuntu/actions-runner/_work/opensearch-operator/opensearch-operator/.tox/integration/lib/python3.10/site-packages/pluggy/_callers.py", line 103, in _multicall
    res = hook_impl.function(*args)
  File "/home/ubuntu/actions-runner/_work/opensearch-operator/opensearch-operator/.tox/integration/lib/python3.10/site-packages/_pytest/python.py", line 159, in pytest_pyfunc_call
    result = testfunction(**testargs)
  File "/home/ubuntu/actions-runner/_work/opensearch-operator/opensearch-operator/.tox/integration/lib/python3.10/site-packages/pytest_asyncio/plugin.py", line 529, in inner
    _loop.run_until_complete(task)
  File "/usr/lib/python3.10/asyncio/base_events.py", line 649, in run_until_complete
    return future.result()
  File "/home/ubuntu/actions-runner/_work/opensearch-operator/opensearch-operator/tests/integration/plugins/test_plugins.py", line 304, in test_monitoring_user_fetch_prometheus_data
    assert response_str.count("opensearch_") > 500
AssertionError: assert 0 > 500
 +  where 0 = <built-in method count of str object at 0x79c198ced030>('opensearch_')
 +    where <built-in method count of str object at 0x79c198ced030> = 'Unauthorized'.count
```

--- a/tests/integration/plugins/test_plugins.py
+++ b/tests/integration/plugins/test_plugins.py
@@ -82,7 +82,7 @@ async def _set_config(ops_test: OpsTest, deploy_type: str, conf: dict[str, str])
     await ops_test.model.applications[APP_NAME].set_config(conf)
 
 
-async def _wait_for_units(ops_test: OpsTest, deployment_type: str) -> None:
+async def _wait_for_units(ops_test: OpsTest, deployment_type: str, extra_apps=[]) -> None:
     """Wait for all units to be active.
 
     This wait will behavior accordingly to small/large.
@@ -90,10 +90,9 @@ async def _wait_for_units(ops_test: OpsTest, deployment_type: str) -> None:
     if deployment_type == "small_deployment":
         await wait_until(
             ops_test,
-            apps=[APP_NAME],
+            apps=[APP_NAME] + extra_apps,
             apps_statuses=["active"],
             units_statuses=["active"],
-            wait_for_exact_units={APP_NAME: 3},
             timeout=1800,
             idle_period=IDLE_PERIOD,
         )
@@ -105,15 +104,10 @@ async def _wait_for_units(ops_test: OpsTest, deployment_type: str) -> None:
             MAIN_ORCHESTRATOR_NAME,
             FAILOVER_ORCHESTRATOR_NAME,
             APP_NAME,
-        ],
+        ]
+        + extra_apps,
         apps_statuses=["active"],
         units_statuses=["active"],
-        wait_for_exact_units={
-            TLS_CERTIFICATES_APP_NAME: 1,
-            MAIN_ORCHESTRATOR_NAME: 1,
-            FAILOVER_ORCHESTRATOR_NAME: 2,
-            APP_NAME: 1,
-        },
         timeout=1800,
         idle_period=IDLE_PERIOD,
     )
@@ -173,7 +167,7 @@ async def test_prometheus_exporter_enabled_by_default(ops_test, deploy_type: str
 async def test_small_deployments_prometheus_exporter_cos_relation(ops_test, deploy_type: str):
     await ops_test.model.deploy(COS_APP_NAME, channel="edge"),
     await ops_test.model.integrate(APP_NAME, COS_APP_NAME)
-    await _wait_for_units(ops_test, deploy_type)
+    await _wait_for_units(ops_test, deploy_type, extra_apps=[COS_APP_NAME])
 
     # Check that the correct settings were successfully communicated to grafana-agent
     cos_leader_id = await get_leader_unit_id(ops_test, COS_APP_NAME)
@@ -263,7 +257,7 @@ async def test_large_deployment_prometheus_exporter_cos_relation(ops_test, deplo
     await ops_test.model.integrate(MAIN_ORCHESTRATOR_NAME, COS_APP_NAME)
     await ops_test.model.integrate(APP_NAME, COS_APP_NAME)
 
-    await _wait_for_units(ops_test, deploy_type)
+    await _wait_for_units(ops_test, deploy_type, extra_apps=[COS_APP_NAME])
 
     leader_id = await get_leader_unit_id(ops_test, APP_NAME)
     leader_name = f"{APP_NAME}/{leader_id}"
@@ -315,7 +309,7 @@ async def test_prometheus_monitor_user_password_change(ops_test, deploy_type: st
     result1 = await run_action(
         ops_test, leader_id, "set-password", {"username": "monitor"}, app=app
     )
-    await _wait_for_units(ops_test, deploy_type)
+    await _wait_for_units(ops_test, deploy_type, extra_apps=[COS_APP_NAME])
 
     new_password = result1.response.get("monitor-password")
     # Now, we compare the change in the action above with the opensearch's nodes.

--- a/tests/integration/plugins/test_plugins.py
+++ b/tests/integration/plugins/test_plugins.py
@@ -94,6 +94,7 @@ async def _wait_for_units(ops_test: OpsTest, deployment_type: str, wait_for_cos:
             apps_statuses=["active"],
             units_statuses=["active"],
             timeout=1800,
+            wait_for_exact_units={APP_NAME: 3},
             idle_period=IDLE_PERIOD,
         )
         if wait_for_cos:
@@ -104,7 +105,7 @@ async def _wait_for_units(ops_test: OpsTest, deployment_type: str, wait_for_cos:
                 timeout=1800,
                 idle_period=IDLE_PERIOD,
             )
-            return
+        return
     await wait_until(
         ops_test,
         apps=[
@@ -113,6 +114,12 @@ async def _wait_for_units(ops_test: OpsTest, deployment_type: str, wait_for_cos:
             FAILOVER_ORCHESTRATOR_NAME,
             APP_NAME,
         ],
+        wait_for_exact_units={
+            TLS_CERTIFICATES_APP_NAME: 1,
+            MAIN_ORCHESTRATOR_NAME: 1,
+            FAILOVER_ORCHESTRATOR_NAME: 2,
+            APP_NAME: 1,
+        },
         apps_statuses=["active"],
         units_statuses=["active"],
         timeout=1800,

--- a/tests/integration/plugins/test_plugins.py
+++ b/tests/integration/plugins/test_plugins.py
@@ -83,7 +83,9 @@ async def _set_config(ops_test: OpsTest, deploy_type: str, conf: dict[str, str])
 
 
 async def _wait_for_units(
-    ops_test: OpsTest, deployment_type: str, wait_for_cos: bool = False,
+    ops_test: OpsTest,
+    deployment_type: str,
+    wait_for_cos: bool = False,
 ) -> None:
     """Wait for all units to be active.
 

--- a/tests/integration/plugins/test_plugins.py
+++ b/tests/integration/plugins/test_plugins.py
@@ -82,7 +82,7 @@ async def _set_config(ops_test: OpsTest, deploy_type: str, conf: dict[str, str])
     await ops_test.model.applications[APP_NAME].set_config(conf)
 
 
-async def _wait_for_units(ops_test: OpsTest, deployment_type: str, extra_apps=[]) -> None:
+async def _wait_for_units(ops_test: OpsTest, deployment_type: str, wait_for_cos=False) -> None:
     """Wait for all units to be active.
 
     This wait will behavior accordingly to small/large.
@@ -90,13 +90,22 @@ async def _wait_for_units(ops_test: OpsTest, deployment_type: str, extra_apps=[]
     if deployment_type == "small_deployment":
         await wait_until(
             ops_test,
-            apps=[APP_NAME] + extra_apps,
+            apps=[APP_NAME],
             apps_statuses=["active"],
             units_statuses=["active"],
             timeout=1800,
             idle_period=IDLE_PERIOD,
         )
-        return
+        if wait_for_cos:
+            await wait_until(
+                ops_test,
+                apps=[COS_APP_NAME],
+                apps_statuses=["active", "blocked"],
+                units_statuses=["active", "blocked"],
+                timeout=1800,
+                idle_period=IDLE_PERIOD,
+            )
+            return
     await wait_until(
         ops_test,
         apps=[
@@ -104,13 +113,21 @@ async def _wait_for_units(ops_test: OpsTest, deployment_type: str, extra_apps=[]
             MAIN_ORCHESTRATOR_NAME,
             FAILOVER_ORCHESTRATOR_NAME,
             APP_NAME,
-        ]
-        + extra_apps,
+        ],
         apps_statuses=["active"],
         units_statuses=["active"],
         timeout=1800,
         idle_period=IDLE_PERIOD,
     )
+    if wait_for_cos:
+        await wait_until(
+            ops_test,
+            apps=[COS_APP_NAME],
+            apps_statuses=["active", "blocked"],
+            units_statuses=["active", "blocked"],
+            timeout=1800,
+            idle_period=IDLE_PERIOD,
+        )
 
 
 @pytest.mark.parametrize("deploy_type", SMALL_DEPLOYMENTS)
@@ -167,7 +184,7 @@ async def test_prometheus_exporter_enabled_by_default(ops_test, deploy_type: str
 async def test_small_deployments_prometheus_exporter_cos_relation(ops_test, deploy_type: str):
     await ops_test.model.deploy(COS_APP_NAME, channel="edge"),
     await ops_test.model.integrate(APP_NAME, COS_APP_NAME)
-    await _wait_for_units(ops_test, deploy_type, extra_apps=[COS_APP_NAME])
+    await _wait_for_units(ops_test, deploy_type, wait_for_cos=True)
 
     # Check that the correct settings were successfully communicated to grafana-agent
     cos_leader_id = await get_leader_unit_id(ops_test, COS_APP_NAME)
@@ -257,7 +274,7 @@ async def test_large_deployment_prometheus_exporter_cos_relation(ops_test, deplo
     await ops_test.model.integrate(MAIN_ORCHESTRATOR_NAME, COS_APP_NAME)
     await ops_test.model.integrate(APP_NAME, COS_APP_NAME)
 
-    await _wait_for_units(ops_test, deploy_type, extra_apps=[COS_APP_NAME])
+    await _wait_for_units(ops_test, deploy_type, wait_for_cos=True)
 
     leader_id = await get_leader_unit_id(ops_test, APP_NAME)
     leader_name = f"{APP_NAME}/{leader_id}"
@@ -309,7 +326,7 @@ async def test_prometheus_monitor_user_password_change(ops_test, deploy_type: st
     result1 = await run_action(
         ops_test, leader_id, "set-password", {"username": "monitor"}, app=app
     )
-    await _wait_for_units(ops_test, deploy_type, extra_apps=[COS_APP_NAME])
+    await _wait_for_units(ops_test, deploy_type, wait_for_cos=True)
 
     new_password = result1.response.get("monitor-password")
     # Now, we compare the change in the action above with the opensearch's nodes.

--- a/tests/integration/plugins/test_plugins.py
+++ b/tests/integration/plugins/test_plugins.py
@@ -82,7 +82,9 @@ async def _set_config(ops_test: OpsTest, deploy_type: str, conf: dict[str, str])
     await ops_test.model.applications[APP_NAME].set_config(conf)
 
 
-async def _wait_for_units(ops_test: OpsTest, deployment_type: str, wait_for_cos: bool = False) -> None:
+async def _wait_for_units(
+    ops_test: OpsTest, deployment_type: str, wait_for_cos: bool = False,
+) -> None:
     """Wait for all units to be active.
 
     This wait will behavior accordingly to small/large.

--- a/tests/integration/plugins/test_plugins.py
+++ b/tests/integration/plugins/test_plugins.py
@@ -100,8 +100,7 @@ async def _wait_for_units(ops_test: OpsTest, deployment_type: str, wait_for_cos=
             await wait_until(
                 ops_test,
                 apps=[COS_APP_NAME],
-                apps_statuses=["active", "blocked"],
-                units_statuses=["active", "blocked"],
+                units_statuses=["blocked"],
                 timeout=1800,
                 idle_period=IDLE_PERIOD,
             )
@@ -123,8 +122,7 @@ async def _wait_for_units(ops_test: OpsTest, deployment_type: str, wait_for_cos=
         await wait_until(
             ops_test,
             apps=[COS_APP_NAME],
-            apps_statuses=["active", "blocked"],
-            units_statuses=["active", "blocked"],
+            units_statuses=["blocked"],
             timeout=1800,
             idle_period=IDLE_PERIOD,
         )

--- a/tests/integration/plugins/test_plugins.py
+++ b/tests/integration/plugins/test_plugins.py
@@ -82,7 +82,7 @@ async def _set_config(ops_test: OpsTest, deploy_type: str, conf: dict[str, str])
     await ops_test.model.applications[APP_NAME].set_config(conf)
 
 
-async def _wait_for_units(ops_test: OpsTest, deployment_type: str, wait_for_cos=False) -> None:
+async def _wait_for_units(ops_test: OpsTest, deployment_type: str, wait_for_cos: bool = False) -> None:
     """Wait for all units to be active.
 
     This wait will behavior accordingly to small/large.

